File: Year 1 Semester ב/DLS/Algorithms/Algorithms.hs
module Algorithms
  ( Tree(..)
  )
where

import           Data.List
import           Data.Maybe

-- Graph Theory

data Graph a = Graph { vertices :: [a]
                   , edges :: [(a, a)]
                   } deriving (Show,Eq)

-- Example:
-- g = Graph {vertices=[0..4],edges=[(0,1),(0,2),(0,3),(1,4),(2,1),(3,2)]}

inEdges :: (Eq a) => Graph a -> a -> [(a, a)]
inEdges (Graph { vertices = v, edges = e }) vert =
  filter (\(x, y) -> y == vert) e

outEdges :: (Eq a) => Graph a -> a -> [(a, a)]
outEdges (Graph { vertices = v, edges = e }) vert =
  filter (\(x, y) -> x == vert) e

sinks :: (Eq a) => Graph a -> [a]
sinks (Graph { vertices = v, edges = e }) =
  let g = Graph { vertices = v, edges = e }
  in  filter ((== 0) . length . (outEdges g)) v

sources :: (Eq a) => Graph a -> [a]
sources (Graph { vertices = v, edges = e }) =
  let g = Graph { vertices = v, edges = e }
  in  filter ((== 0) . length . (inEdges g)) v

-- subgraph :: (Eq a) => Graph a -> [a] -> Graph a
-- subgraph (Graph { vertices = v, edges = e }) vs = Graph { vertices = filter (`elem` vs) v
-- , edges    = filter (\x -> not $ elem x $ inEdges g s) e

topOrd :: (Eq a) => Graph a -> [Int]
topOrd (Graph { vertices = v, edges = e }) = if length v == 1
  then [0]
  else
    let g   = Graph { vertices = v, edges = e }
        s   = (sinks g) !! 0
        tpl = splitAt (fromJust $ elemIndex s v) $ topOrd
          (Graph { vertices = filter (/= s) v
                 , edges    = filter (\x -> not $ elem x $ inEdges g s) e
                 }
          )
    in  fst tpl ++ [length v - 1] ++ snd tpl

sortFromOrd :: (Eq a) => Graph a -> a -> Int
sortFromOrd g =
  let ts = topOrd g in \x -> ts !! (fromJust $ elemIndex x $ vertices g)

topSort :: (Eq a) => Graph a -> Graph Int
topSort g =
  let conv = sortFromOrd g
  in  Graph { vertices = [0 .. (length (vertices g) - 1)]
            , edges    = map (\(x, y) -> (conv x, conv y)) $ edges g
            }

delay :: (Eq a) => Graph a -> [(a, Int)]
delay g =
  let
    conv = sortFromOrd g
    ts   = topSort g
    d =
      foldl
          (\acc j ->
            acc
              ++ [ if length (inEdges ts j) /= 0
                     then
                       1 + maximum
                         [ acc !! i | (i, k) <- edges ts, k == j, i < j ]
                     else 0
                 ]
          )
          []
        $ vertices ts
  in
    map (\x -> (x, d !! (conv x))) $ vertices g

-- Boolean Formula

data Tree a = Node a | Branch a [Tree a] deriving (Eq)

-- makeTree :: (Eq a) => Graph a -> Maybe (Tree a)
-- makeTree g
--   | ((/= 1) . length . sinks $ g)
--     || (v /= filter ((`elem` [0, 1]) . length . outEdges g) v)
--   = Nothing
--   | otherwise
--   = Just $ makeTree' g
--   where v = vertices g

-- -- g = Graph {vertices=[0..4],edges=[(1,0),(2,0),(3,1),(4,1)]}

-- makeTree' :: (Eq a) => Graph a -> Tree a
-- makeTree' (Graph { vertices = [x], edges = [] }) = Node x
-- makeTree' g =
--   let r = (sources g) !! 0 in Branch r ((map makeTree' . hangingTree g) r)

-- hangingTree :: (Eq a) => Graph a -> a -> [Graph a]
-- hangingTree g r
--   | e == [] = []
--   | otherwise = joinGraph
--     (foldl joinGraph . filter (/= []) . map (hangingTree g . fst) $ e)
--     (Graph { vertices = [r], edges = inEdges g r })
--   where e = inEdges g r

show' :: (Show a) => a -> String
show' = (filter (not . (`elem` ['\'', '\"']))) . show

-- INORDER

instance (Show a) => Show (Tree a) where
  show (Node a       ) = show' a
  show (Branch op [a]) = "(" ++ show' op ++ show' a ++ ")"
  show (Branch op [a, b]) =
    "(" ++ show' a ++ " " ++ show' op ++ " " ++ show' b ++ ")"
  show (Branch op lst) =
    "(" ++ show' op ++ " (" ++ (intercalate ", " (map show' lst)) ++ "))"

-- t = Branch "AND" [Branch "OR" [Node "X", Node "0"], Branch "~" [Node "Y"]]

File: Year 1 Semester ב/DLS/Algorithms/Algorithms.py
import functools as f


def deg(sign, v, E):
    return (
        len([e for e in E if e[1] == v])
        if sign == 1  # in
        else len([e for e in E if e[0] == v])
        if sign == -1  # out
        else 0
    )


def sink(V, E):
    return [v for v in V if deg(-1, v, E) == 0][0]


def subgraph(V, E, U):
    return ([v for v in V if v in U], [(u, v) for u, v in E if u in U and v in U])


def minus_graph(V, E, U_):
    return (
        [v for v in V if v not in U_],
        [(u, v) for u, v in E if u not in U_ and v not in U_],
    )


def topOrd(V, E):
    if len(V) == 1:
        return [0]
    else:
        s = sink(V, E)
        i = V.index(s)
        pi = topOrd(*minus_graph(V, E, [s]))
        return pi[:i] + [len(V) - 1] + pi[i:]


def topSort(V, E, pi=None):
    if pi == None:
        pi = topOrd(V, E)
    return list(range(len(V))), [(pi[V.index(x)], pi[V.index(y)]) for x, y in E]


G = ([0, 1, 2, 3, 4], [(0, 1), (0, 2), (0, 3), (1, 4), (2, 1), (3, 2)])
print(topOrd(*G))
print(*topSort(*G))


def longestPath(V, E, pi=None):
    V_, E_ = topSort(V, E, pi)
    return f.reduce(
        lambda acc, j: acc
        + [
            1 + max(acc[i] for i, k in E_ if k == j and i < j)
            if not deg(1, j, E_) == 0
            else 0
        ],
        V_,
        [],
    )


def delay(V, E):
    pi = topOrd(V, E)
    d = longestPath(V, E, pi)
    return {v: d[pi[V.index(v)]] for v in V}


print(longestPath(*G))
print(delay(*G))

print("-" * 40)

R = ([0, 1, 2, 3, 4, 5], [(1, 0), (2, 0), (3, 1), (4, 1), (5, 2)])

print(topOrd(*R))
print(*topSort(*R))
print(longestPath(*R))
print(delay(*R))


def decompose(V, E):
    r = sink(V, E)
    return [minus_graph(V, E, [r_i]) for r_i, v in E if v == r]

